### ISSUE TITLE
`on_failure` causing sorbet to lose type information in initializer

### ISSUE THREAD

**Issue by jeffcarbs** [sorbet/sorbet#2970]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(x%3A%20Integer).void.on_failure%7D%0A%20%20def%20initialize(x)%0A%20%20%20%20%40x%20%3D%20x%0A%20%20end%0A%20%20%0A%20%20def%20bar%0A%20%20%20%20T.reveal_type(%40x)%0A%20%20%20%20%40x.upcase%0A%20%20end%0Aend%0A)

```ruby
# typed: true
class A
  extend T::Sig

  sig {params(x: Integer).void.on_failure}
  def initialize(x)
    @x = x
  end
  
  def bar
    T.reveal_type(@x)
    @x.upcase
  end
end
```

#### Observed output

```
editor.rb:11: Revealed type: T.untyped https://srb.help/7014
    11 |    T.reveal_type(@x)
            ^^^^^^^^^^^^^^^^^
  From:
Errors: 1
```

#### Expected behavior

I would expect
- Revealed type should be `Integer`
- Error raised on line 12 saying `Method upcase does not exist on Integer`

To reproduce expected behavior, remove the `on_failure` from the sig on the `initialize` method.

- - -

**Comment by elliottt** [comment]
The problem is with [`ASTUtil::castSig`](https://github.com/sorbet/sorbet/blob/634cc2d0a2594383612430c1ef1c452101725bd2/rewriter/Util.cc#L168), which requires that sigs end with `returns` (or `void`). A workaround is to move `on_failure` before `void`, so that your program looks like this instead:

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(x%3A%20Integer).on_failure.void%7D%0A%20%20def%20initialize(x)%0A%20%20%20%20%40x%20%3D%20x%0A%20%20end%0A%20%20%0A%20%20def%20bar%0A%20%20%20%20T.reveal_type(%40x)%0A%20%20%20%20%40x.upcase%0A%20%20end%0Aend)

```ruby
# typed: true
class A
  extend T::Sig

  sig {params(x: Integer).on_failure.void}
  def initialize(x)
    @x = x
  end
  
  def bar
    T.reveal_type(@x)
    @x.upcase
  end
end
```

**Comment by connorshea** [comment]
Could we just make Sorbet fail typechecking statically if you have anything after the void/returns?

**Comment by elliottt** [comment]
I'd prefer if we just fixed [ASTUtil::castSig](https://github.com/sorbet/sorbet/blob/634cc2d0a2594383612430c1ef1c452101725bd2/rewriter/Util.cc#L168), as that would lead to less surprising behavior.

**Comment by jez** [comment]
Fixed Fixed in #3159
